stellargraph/layer/deep_graph_infomax.py

```diff
@@ -0,0 +1,156 @@
+
+
+@experimental(reason="lack of unit tests", issues=[1003])
+class DeepGraphInfoMax:
+    """
+    A class to wrap stellargraph models for Deep Graph Infomax unsupervised training
+    (https://arxiv.org/pdf/1809.10341.pdf).
+
+    Args:
+        base_model: the base stellargraph model class
+    """
+
+    def __init__(self, base_model):
+
+        if not isinstance(base_model, (GCN, GAT, APPNP, PPNP)):
+            raise TypeError(
+                f"base_model: expected GCN, GAT, APPNP or PPNP found {type(base_model).__name__}"
+            )
+
+        self.base_model = base_model
+
+        self._node_feats = None
+
+        # specific to full batch models
+        self._corruptible_inputs_idxs = [0]
+
+    def build(self):
+        """
+        A function to create the the keras inputs and outputs for a Deep Graph Infomax model for unsupervised training.
+
+        Note that the tf.nn.sigmoid_cross_entropy_with_logits loss must be used with this model.
+
+        Example::
+
+            dg_infomax = DeepGraphInfoMax(...)
+            x_in, x_out = dg_infomax.build()
+            model = Model(inputs=x_in, outputs=x_out)
+            model.compile(loss=tf.nn.sigmoid_cross_entropy_with_logits, ...)
+
+        Returns:
+            input and output layers for use with a keras model
+        """
+
+        if self.base_model.multiplicity != 1:
+            warnings.warn(
+                f"multiplicity: expected the base_model to have a multiplicity of 1, found"
+                f" ({self.base_model.multiplicity}). A multiplicity of 1 will be used to construct the base model."
+            )
+
+        x_inp, self._node_feats = self.base_model.build(multiplicity=1)
+        x_corr = [
+            Input(batch_shape=x_inp[i].shape) for i in self._corruptible_inputs_idxs
+        ]
+
+        # shallow copy normal inputs and replace corruptible inputs with new inputs
+        x_in_corr = x_inp.copy()
+        for i, x in zip(self._corruptible_inputs_idxs, x_corr):
+            x_in_corr[i] = x
+
+        node_feats_corr = self.base_model(x_in_corr)
+
+        summary = tf.keras.activations.sigmoid(
+            GlobalAveragePooling1D()(self._node_feats)
+        )
+
+        discriminator = BilinearTransform()
+        scores = discriminator([self._node_feats, summary])
+        scores_corrupted = discriminator([node_feats_corr, summary])
+
+        x_out = tf.concat([scores, scores_corrupted], axis=2)
+
+        return x_corr + x_inp, x_out
+
+    def embedding_model(self, model):
+        """
+        A function to create the the inputs and outputs for an embedding model.
+
+        Args:
+            model (keras.Model): the base Deep Graph Infomax model with inputs and outputs created from
+                DeepGraphInfoMax.build()
+        Returns:
+            input and output layers for use with a keras model
+        """
+        x_emb_in = model.inputs[len(self._corruptible_inputs_idxs) :]
+
+        x_emb_out = self._node_feats
```

Comment:
What happens if `embedding_model` is called before `build`?

Reply:
`x_emb_out` will be `None` and the user will get a (probably) confusing error message when they try to buildtheir `keras.Model`. Same outcome if they call `DeepGraphInfoMax.embedding_model` with a `model` not created by the right instance of `DeepGraphInfoMax`. 

I'll catch both of these cases here and give an understandable error message